libraries/botbuilder-ai/src/qnaMakerDialog.ts

```diff
@@ -0,0 +1,359 @@
+        this.threshold = threshold;
+        this.top = top;
+        this.activeLearningCardTitle = activeLearningCardTitle;
+        this.cardNoMatchText = cardNoMatchText;
+        this.strintFilters = strictFilters;
+        this.noAnswer = noAnswer;
+        this.cardNoMatchResponse = cardNoMatchResponse;
+
+        this.addStep(this.callGenerateAnswer.bind(this));
+        this.addStep(this.callTrain.bind(this));
+        this.addStep(this.checkForMultiTurnPrompt.bind(this));
+        this.addStep(this.displayQnAResult.bind(this));
+    }
+
+    public async beginDialog(dc: DialogContext, options?: object): Promise<DialogTurnResult> {
+        if (!dc) { throw new Error('Missing DialogContext'); }
+        
+        if (dc.context.activity.type != ActivityTypes.Message) 
+        { 
+            return Dialog.EndOfTurn; 
+        }
+
+        const dialogOptions: QnAMakerDialogOptions = {
+            qnaDialogResponseOptions: await this.getQnAResponseOptions(),
+            qnaMakerOptions: await this.getQnAMakerOptions()
+        };
+
+        if(options)
+        {
+            Object.assign(dialogOptions, options);
+        }
+
+        return await super.beginDialog(dc, dialogOptions);
+    }
+
+    /**
+     * Returns a new instance of QnAMakerOptions.
+    **/
+    private async getQnAMakerOptions(): Promise<QnAMakerOptions> {
+        return {
+            scoreThreshold: this.threshold,
+            strictFilters: this.strintFilters,
+            top: this.top,
+            qnaId: 0,
+            rankerType: RankerTypes.default,
+            isTest: false
+        };
+    };
+    
+    /**
+     * Returns a new instance of QnAMakerResponseOptions.
+    **/
+    private async getQnAResponseOptions(): Promise<QnAMakerDialogResponseOptions> {
+        return {
+            activeLearningCardTitle: this.activeLearningCardTitle,
+            cardNoMatchResponse: this.cardNoMatchResponse,
+            cardNoMatchText: this.cardNoMatchText,
+            noAnswer: this.noAnswer
+        };
+    }
+
+    /**
+     * Queries the knowledgebase and either passes result to the next step or constructs and displays an active learning card
+     * if active learning is enabled and multiple score close answers are returned.
+    **/
+    private async callGenerateAnswer(step: WaterfallStepContext): Promise<DialogTurnResult> {
+        const dialogOptions: QnAMakerDialogOptions = step.activeDialog.state[this.options];
+        dialogOptions.qnaMakerOptions.qnaId = 0;
+        dialogOptions.qnaMakerOptions.context = { previousQnAId: 0, previousUserQuery: '' };
+
+        step.values[this.currentQuery] = step.context.activity.text;
+        const previousContextData: { [key: string]: number } = step.activeDialog.state[this.qnAContextData] ? step.activeDialog.state[this.qnAContextData] : {};
+        var previousQnAId = step.activeDialog.state[this.previousQnAId] ? step.activeDialog.state[this.previousQnAId] : 0;
```

Comment:
>step.activeDialog.state[this.previousQnAId] ? step.activeDialog.state[this.previousQnAId] : 0 [](start = 28, length = 93)

You can simplify this to step.activeDialog.state[this.previousQnAId] || 0;

Reply:
Thanks - altered this in a few places where I was using the longer hand. Good tip!